### DISCUSSION TITLE
support recursive anonymous function

### DISCUSSION THREAD

**Discussion by SleepyRoy** [vlang/v#11665] [Ideas]
**Update: A preliminary decision is to accept this. Syntax suggestions are wanted and welcome!**

Local functions are very useful. Thanks to the recently implemented anonymous function, we can now write, e.g.,
```V
fn func() {
    local := fn() {
        println('hello world')
    }
    local()
}
```
However, sometimes local functions need to be recursive, so please consider implementing recursive anonymous function. Here's a possible syntax.
```V
fn func() {
    local := fn(n int) {
        if n>1 {
            fn(n-1)
        }
        println('hello world')
    }
    local(10)
}
```
A more real example:
```V
// example: a binary tree
struct Empty {}
struct Node {value int  left Tree  right Tree}
Type Tree = Empty | Node 

get_size_of_tree := fn (tree Tree) int {
	return match tree {
		Empty { 0 }
		Node { 1 + fn(tree.left) + fn(tree.right) }
	}
}
size := get_size_of_tree(tree)
```

**Comment by ntrel** [comment]
V doesn't allow local fn declarations, but in this case it would be better for the parser than `fn(n-1)`:
```v
    fn local(n int) {
        if n>1 {
            local(n-1)
        }
        println('hello world')
    }
```

**Reply by viocha** [reply]
Why can't local fn declarations be supported? If they were supported, there wouldn't be the problem of being unable to define recursive local functions.

**Reply by larpon** [reply]
Closures and local functions are supported

**Reply by viocha** [reply]
I mean this syntax

```
    fn outer(){
      fn dfs(){
        dfs() // recursively call
      }
    }
 ```

**Reply by larpon** [reply]
What do you need them for?

**Reply by JalonSolov** [reply]
In that case, `dfs()` isn't an anonymous function.  It's a named inner function.  This issue is about supporting recursive anonymous functions.

To support recursive anonymous functions, you do need some way to refer to them... either a name, a placeholder (such as `this`, etc.), or a keyword of some sort to imply the same thing.

**Reply by viocha** [reply]
> What do you need them for?

Since there is a need for local functions and also a need for recursive functions, isn't it natural to have both requirements?

**Reply by JalonSolov** [reply]
Personally, I'd rather just have a named recursive function.  Why does it need to be anonymous?

**Reply by larpon** [reply]
> > What do you need them for?
> 
> Since there is a need for local functions and also a need for recursive functions, isn't it natural to have both requirements?

... but what do *you* need them for?

**Reply by viocha** [reply]
> > > What do you need them for?
> > 
> > 
> > Since there is a need for local functions and also a need for recursive functions, isn't it natural to have both requirements?
> 
> ... but what do _you_ need them for?

Creating a recursive anonymous function can cause the function definition to not be found during recursive calls, while creating a local named function should avoid this issue.

**Reply by JalonSolov** [reply]
Why does it need to be a local named function?  Why can't it simply be another function?  A plain ol' recursive function.

Nobody has given a good reason to allow anonymous functions to be recursive, when if you want recursive you can simply call a regular recursive function.  The _only_ reason to try to avoid this is to avoid "polluting the namespace" with a function name that may only be used in one place.

That's a... silly reason.  If you need it twice, you're creating 2 functions, instead of one.  Why do you want to complicate/bloat your code that way?

Supporting a "local named function" is really the same thing.  It would be a way to avoid "polluting" the outer namespace with a function that may only be used in one place.

Again... a silly reason.  Same as before... what if you need it more than once?

Just create a regular recursive function and call it.  That doesn't require any changes to the compiler.

Now, if you can come up with a _good_ reason either option should be allowed (not "because it would be cool" or "because I want it"), considering the amount of work it will require to support either one, then please provide said reason.

**Reply by viocha** [reply]
I know you mean that it can be easily abused and thus cause problems. For the problem you mentioned, its source is not supporting recursive calls, but rather that local functions themselves may be abused. In fact, other syntaxes may also be abused. The quality of code still depends on the individual developer. Even a language as concise as Go may produce "spaghetti code".

Most languages that support local functions (such as Rust,C#, Go, Java, C++,JavaScript etc.) usually allow recursive calls because recursion is a natural programming paradigm. Providing the option of local recursive functions does not bring too many problems.

In my opinion, as long as local recursive functions are used reasonably, global scope pollution can be avoided. Closures can also be used to reduce the number of function parameters in function declarations and recursive calls, which can increase code readability. It also makes the code logic more compact and easier to maintain.

**Reply by mindplay-dk** [reply]
Did you see my reply [here](https://github.com/vlang/v/discussions/11665#discussioncomment-1365753)?

The ability to to simply name an anonymous function would be nice to have either way - it can add a touch of semantics for the human reading the code, and might could be used to improve error messages as well.

There's something to be said for just having consistent function syntax anyway - not that javascript is a prime example of great language design, but I have a personal preference for the long form `function` keyword just because it lets me name the function.

The following works in JS and does what you think:

```js
const f = function countdown(n) {
  console.log(n--)
  if (n > 0) {
    countdown(n)
  }
}

f(10)
```

I rarely use recursion, but I'd say, what someone chooses to do (or not do) with that feature is up to them. ☺️

**Reply by JalonSolov** [reply]
A named function is not an anonymous function, so naming them defeats the purpose of them.

**Reply by viocha** [reply]
> A named function is not an anonymous function, so naming them defeats the purpose of them.

Maybe it would be more concise and reduce the learning curve if local functions could follow the same syntax as global functions.

**Reply by mindplay-dk** [reply]
> A named function is not an anonymous function, so naming them defeats the purpose of them.

The value of first-class functions is they can be treated like variables - can be passed as an argument to another function, can be returned by another function, can be assigned as a value to a variable, and so on.

Whether or not a function has a name is just a detail - if it's optional, and if the generated symbol is introduced into the local scope, having a name is completely benign, it's just an option if you'd like to reference it locally (e.g. for recursion) or if the name helps improve your understanding of what the function is or does. (in that context, when referenced somewhere else, or in error messages.)

**Reply by JalonSolov** [reply]
V doesn't support inner named functions, only inner anonymous functions.

If you open it to allow for inner named functions, why does the inner function have to only be called from it's parent?  Why not allow `fn_parent.fn_child.fn_grandchild()`?  Mainly because you're now getting into the realm of insanity... or OOP.

Either way, the first thing V claims is `Simple`... increasing the complexity of the language by allowing inner named functions is far from simple.  It will make the code harder to reason about, as well as requiring a lot of changes to the parser/checker/codegen.

Just make a regular function and call it.  This functionality is already in place.  No changes necessary.

I still haven't seen anyone suggest a good reason why V _needs_ either recursive anonymous functions, _or_ inner named functions.  Simply because something _could_ be done does not mean it _should_ be done.

**Reply by viocha** [reply]
> V doesn't support inner named functions, only inner anonymous functions.
> 
> If you open it to allow for inner named functions, why does the inner function have to only be called from it's parent? Why not allow `fn_parent.fn_child.fn_grandchild()`? Mainly because you're now getting into the realm of insanity... or OOP.
> 
> Either way, the first thing V claims is `Simple`... increasing the complexity of the language by allowing inner named functions is far from simple. It will make the code harder to reason about, as well as requiring a lot of changes to the parser/checker/codegen.
> 
> Just make a regular function and call it. This functionality is already in place. No changes necessary.
> 
> I still haven't seen anyone suggest a good reason why V _needs_ either recursive anonymous functions, _or_ inner named functions. Simply because something _could_ be done does not mean it _should_ be done.

All local functions in all other languages support recursion, only you are being stubborn. I also don't see a valid reason why you would create a function that doesn't support recursion.

**Reply by JalonSolov** [reply]
"All other languages do it" is still not a valid reason.

**Reply by mindplay-dk** [reply]
You see posts like [this](https://github.com/vlang/v/discussions/11665#discussioncomment-1365748) though?

"but it worked there, why doesn't it work here?"

Why *shouldn't* first class functions have features and syntax consistent with regular functions? Why are there two kinds of functions in the first place? And then you need to have some expert come in and explain the intricacies and details of *why* - stuff that doesn't help that person be more productive, just things they need to remember and work around.

Consistency is something to strive for in any programming language. You can argue it adds complexity to the *implementation*, but in this case I'd argue it simplifies the language - being more consistent means fewer surprises, less learning curve.

**Reply by JalonSolov** [reply]
Anonymous functions are an adjunct to V.  They were never planned originally.  Then people wanted closures, so those were added.  Then they wanted anonymous functions, and those were added.

Only enough was added to make each one work... V was not redesigned at those points.

It would take a _large_ amount of effort to go back and redesign V to treat anonymous functions the same as regular functions, allowing recursion, etc.

That's what I've been trying to convey all along... this late in the game, can you give a single, solid reason why V should go through that much upheaval to support this small bit of functionality, when you can already get the same thing just by creating a regular function?

One of V's main tenets, and the first one listed on the home page is "Simple".  It is simple to create a regular function.  It is not simple to go back and redesign parts of the compiler to add functionality that is already covered by the simple functionality that is already there.

**Reply by Wajinn** [reply]
In this case, in agreement with JalonSolov. It's understandable for team V to be very careful about adding features, at this point. Especially those that would be difficult to implement, add excessive complexity, or create stability issues.

**Reply by mindplay-dk** [reply]
Yeah, that's totally fair. I'm arguing the point as a language consumer. I have no idea what it's like to work on this codebase. ☺️

**Comment by SleepyRoy** [comment]
@ntrel but I remember local fn declaration examples like this worked...
```V
fn func() {
    local := fn() {
        println('hello world')
    }
    local()
}
```

**Comment by ntrel** [comment]
>     local := fn() {

That works (I meant you can't have named function declarations inside a function). In the body of the anonymous function above, you can't use the name `local` because it hasn't been defined yet. Anyway, another solution would be to have a special name like `self`: `self(n-1)`.

**Comment by SleepyRoy** [comment]
@ntrel yes that makes sense. I think we could simply use `fn` as such a special name.

**Comment by ntrel** [comment]
`fn(ident)` looks like a function type to the parser. It's best to avoid syntax that requires special tests to tell the difference.

**Comment by SleepyRoy** [comment]
**Update: A preliminary decision is to accept this. Syntax suggestions are wanted and welcome!**

**Comment by mindplay-dk** [comment]
Having calls like `fn(x)` is not very semantic.

And what happens if you have functions within functions and need to call the outermost `fn`?

My suggestion would be an optional means of declaring a symbol, e.g.:

```
// example: a binary tree
struct Empty {}
struct Node {value int  left Tree  right Tree}
Type Tree = Empty | Node 

get_size_of_tree := fn count_nodes(tree Tree) int {
	return match tree {
		Empty { 0 }
		Node { 1 + count_nodes(tree.left) + count_nodes(tree.right) }
	}
}
size := get_size_of_tree(tree)
```

The symbol `count_nodes` in this example would be locally scoped, and calls like `count_nodes(tree.left)` actually look and sound like what they are.

> `fn(ident)` looks like a function type to the parser.

It also looks ambiguous to a person.

Recursive functions in other langs typically need a name, e.g. in JavaScript - so this wouldn't be surprising to most people.

This seems more consistent with function syntax/semantics (in V) in general.
